### DISCUSSION TITLE
Adding modules to a cypress container

### DISCUSSION THREAD

**Discussion by Obelix74Git** [cypress-io/cypress#29284] [Questions and Help]
Hi,
We use the container cypress/included to run our tests. In our test scripts, the cypress developers use the module cypress-mochawesome-reporter. However, this module is not installed in the container cypress/included and I therefore have to add it myself.
How do you add a module to a container, and to make it even more challenging, in an environment that has no internet connectivity?
Thanks for your help.

**Comment by portyu9** [comment]
Hi @Obelix74Git 

The usual approach is to build your own image **from** `cypress/included` and install the additional npm dependencies into that image:

```dockerfile
FROM cypress/included:<fixed-version-tag>

WORKDIR /e2e

COPY package*.json ./

# Cypress is already installed in cypress/included
ENV CYPRESS_INSTALL_BINARY=0

RUN npm ci

COPY . .
```

Add `cypress-mochawesome-reporter` to your project's dependencies before building the image:

```bash
npm install --save-dev cypress-mochawesome-reporter
```

For an environment with **no internet access**, the important part is to build this image somewhere that does have access and then transfer it to the offline environment, for example through an internal Docker registry or `docker save` / `docker load`.

If the Docker build itself must also be offline, you will need an internal npm registry/cache or otherwise provide `cypress-mochawesome-reporter` and all of its transitive npm dependencies locally. Docker/Cypress cannot install a package that is not available inside that environment.

Also prefer a **fixed `cypress/included` image tag** instead of `latest` so the image does not change unexpectedly.

Please mark this answer as accepted if it helped, thank you.
